### PR TITLE
add metrics sinks benchmarking

### PR DESCRIPTION
**Pull request by amoore877**
Signed-off-by: Andrew Moore <andrew.s.moore@uber.com>

**Pull Request check list**

- [X] Commit conforms to CONTRIBUTING.md?
- [X] Proper tests/regressions included?
- [X] Documentation updated?

**Affected functionality**
N/A

**Description of change**
Investigative `Metrics` benchmarking related to #1040 and questions about `InMem` performance arising from #1248. Will post current results as comment to this PR.
Minor nit fix of `testConfig` function for Prometheus -> should be `testPrometheusConfig` to match other similar functions for the different metrics sinks.

**Which issue this PR fixes**
N/A

### CODE REVIEW

**Comment by amoore877** [comment]
```
goos: darwin
goarch: amd64
pkg: github.com/spiffe/spire/pkg/common/telemetry
BenchmarkDogStatsd/SetGauge-8         	  263077	      4543 ns/op	     336 B/op	       9 allocs/op
BenchmarkDogStatsd/SetGaugeWithLabels-8         	  174104	      7497 ns/op	    1066 B/op	      32 allocs/op
BenchmarkDogStatsd/EmitKey-8                    	 7159405	       177 ns/op	      96 B/op	       1 allocs/op
BenchmarkDogStatsd/IncrCounter-8                	  321697	      3886 ns/op	     336 B/op	       9 allocs/op
BenchmarkDogStatsd/IncrCounterWithLabels-8      	  185707	      6912 ns/op	    1045 B/op	      32 allocs/op
BenchmarkDogStatsd/AddSample-8                  	  256438	      4336 ns/op	     336 B/op	       9 allocs/op
BenchmarkDogStatsd/AddSampleWithLabels-8        	  171464	      6949 ns/op	    1063 B/op	      32 allocs/op
BenchmarkDogStatsd/MeasureSince-8               	  251042	      4662 ns/op	     352 B/op	       9 allocs/op
BenchmarkDogStatsd/MeasureSinceWithLabels-8     	  159582	      7339 ns/op	    1067 B/op	      32 allocs/op
BenchmarkInMem/SetGauge-8                       	  701534	      1782 ns/op	     872 B/op	      21 allocs/op
BenchmarkInMem/SetGaugeWithLabels-8             	  297966	      4122 ns/op	    1711 B/op	      45 allocs/op
BenchmarkInMem/EmitKey-8                        	  935707	      1261 ns/op	     638 B/op	      14 allocs/op
BenchmarkInMem/IncrCounter-8                    	  698413	      1784 ns/op	     872 B/op	      21 allocs/op
BenchmarkInMem/IncrCounterWithLabels-8          	  309618	      4123 ns/op	    1685 B/op	      45 allocs/op
BenchmarkInMem/AddSample-8                      	  747271	      1723 ns/op	     872 B/op	      21 allocs/op
BenchmarkInMem/AddSampleWithLabels-8            	  309001	      3967 ns/op	    1685 B/op	      45 allocs/op
BenchmarkInMem/MeasureSince-8                   	  691434	      1851 ns/op	     872 B/op	      21 allocs/op
BenchmarkInMem/MeasureSinceWithLabels-8         	  283728	      4088 ns/op	    1685 B/op	      45 allocs/op
BenchmarkM3/SetGauge-8                          	 1000000	      1143 ns/op	     944 B/op	      11 allocs/op
BenchmarkM3/SetGaugeWithLabels-8                	  419714	      3054 ns/op	    1467 B/op	      26 allocs/op
BenchmarkM3/EmitKey-8                           	 7059409	       171 ns/op	      96 B/op	       1 allocs/op
BenchmarkM3/IncrCounter-8                       	 1000000	      1162 ns/op	     944 B/op	      11 allocs/op
BenchmarkM3/IncrCounterWithLabels-8             	  407708	      2943 ns/op	    1466 B/op	      26 allocs/op
BenchmarkM3/AddSample-8                         	 1000000	      1160 ns/op	     976 B/op	      12 allocs/op
BenchmarkM3/AddSampleWithLabels-8               	  396505	      3085 ns/op	    1498 B/op	      27 allocs/op
BenchmarkM3/MeasureSince-8                      	  962300	      1262 ns/op	     976 B/op	      12 allocs/op
BenchmarkM3/MeasureSinceWithLabels-8            	  389314	      3145 ns/op	    1500 B/op	      27 allocs/op
BenchmarkPrometheus/SetGauge-8                  	  884250	      1419 ns/op	     419 B/op	      11 allocs/op
BenchmarkPrometheus/SetGaugeWithLabels-8        	  354684	      3844 ns/op	    1097 B/op	      34 allocs/op
BenchmarkPrometheus/EmitKey-8                   	 7216338	       186 ns/op	      96 B/op	       1 allocs/op
BenchmarkPrometheus/IncrCounter-8               	  722590	      1716 ns/op	     419 B/op	      11 allocs/op
BenchmarkPrometheus/IncrCounterWithLabels-8     	  261640	      4151 ns/op	    1097 B/op	      34 allocs/op
BenchmarkPrometheus/AddSample-8                 	  823626	      1554 ns/op	     419 B/op	      11 allocs/op
BenchmarkPrometheus/AddSampleWithLabels-8       	  324879	      3640 ns/op	    1099 B/op	      34 allocs/op
BenchmarkPrometheus/MeasureSince-8              	  821035	      1761 ns/op	     419 B/op	      11 allocs/op
BenchmarkPrometheus/MeasureSinceWithLabels-8    	  318978	      3710 ns/op	    1097 B/op	      34 allocs/op
BenchmarkStatsd/SetGauge-8                      	2019/11/20 14:06:44 [ERR] Error writing to statsd! Err: write udp [::1]:53305->[::1]:8125: write: connection refused
 1000000	      1090 ns/op	     292 B/op	       7 allocs/op
BenchmarkStatsd/SetGaugeWithLabels-8            	  420337	      3009 ns/op	    1008 B/op	      23 allocs/op
BenchmarkStatsd/EmitKey-8                       	 1391508	       854 ns/op	     180 B/op	       5 allocs/op
BenchmarkStatsd/IncrCounter-8                   	2019/11/20 14:06:49 [ERR] Error writing to statsd! Err: write udp [::1]:63148->[::1]:8125: write: connection refused
 1000000	      1098 ns/op	     292 B/op	       7 allocs/op
BenchmarkStatsd/IncrCounterWithLabels-8         	  421587	      2978 ns/op	    1008 B/op	      23 allocs/op
BenchmarkStatsd/AddSample-8                     	 1000000	      1070 ns/op	     292 B/op	       7 allocs/op
BenchmarkStatsd/AddSampleWithLabels-8           	  384312	      2984 ns/op	    1009 B/op	      23 allocs/op
BenchmarkStatsd/MeasureSince-8                  	2019/11/20 14:06:54 [ERR] Error writing to statsd! Err: write udp [::1]:59497->[::1]:8125: write: connection refused
 1000000	      1240 ns/op	     292 B/op	       7 allocs/op
BenchmarkStatsd/MeasureSinceWithLabels-8        	  403268	      3120 ns/op	    1007 B/op	      23 allocs/op
PASS
ok  	github.com/spiffe/spire/pkg/common/telemetry	59.747s
Success: Benchmarks passed.
```

**Comment by amoore877** [comment]
will Travis CI automatically run benchmarking?

**Comment by amoore877** [comment]
unsure on `statsd` errors at this time

**Comment by amoore877** [comment]
Generally it would be relied on for the imported metrics library to have such benchmarking, but this is not the case today.
However, it would be perfectly understandable to not merge based on that philosophy and take this posting as simple documentation for future efforts / evaluations.

File: pkg/common/telemetry/metrics_benchmark_test.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+		for n := 0; n < b.N; n++ {
+			m.MeasureSinceWithLabels(key, valt, labels)
+		}
+	})
+}
+
+func getDogStatsdMetricImpl(b *testing.B) Metrics {
+	m, err := NewMetrics(testDogStatsdConfig())
+	if err != nil {
+		b.Error(err)
```
Comment (azdagron) [inline pkg/common/telemetry/metrics_benchmark_test.go]: b.Error marks the function as failed but does not immediately exit the goroutine (i.e. runs Fail, not FailNow)

Comment (azdagron) [reply pkg/common/telemetry/metrics_benchmark_test.go]: maybe just use `require.NoError(t, err)`?

Comment (amoore877) [reply pkg/common/telemetry/metrics_benchmark_test.go]: can you have both T and B in a single test? Couldn't find anything allowing that. Using 
```
		b.Log(err)
		b.FailNow()
```
for now

still curious about the statsd error

Comment (azdagron) [reply pkg/common/telemetry/metrics_benchmark_test.go]: sorry, `require.NoError(b, err)`. When you create functions that you want to use in both tests and benchmarks, then you usually have use the `testing.TB` interface as a parameter.

The `stretchr/testify` package uses an even more paired down interface (ie. `require.TestingT`) that both `*testing.T` and `*testing.B` conform to. So you can do `require.NoError` within both tests and benchmarks.

**Comment by azdagron** [comment]
The statsd sink writes udp packets to localhost 8125 (the configured stats server). Since you don't have anything bound to that port we log that the metric wasn't able to be written. Maybe `testStatsdConfig` can take a port number (to init the config) and you can start up a localhost udp listener that just discards incoming packets?

File: pkg/common/telemetry/metrics_benchmark_test.go (hunk 1 above)
Comment (amoore877) [reply pkg/common/telemetry/metrics_benchmark_test.go]: very nice :)

File: pkg/common/telemetry/metrics_benchmark_test.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+func BenchmarkPrometheus(b *testing.B) {
+	m := getPrometheusMetricImpl(b)
+
+	benchmarkMetricImpl(b, m)
+}
+
+func BenchmarkStatsd(b *testing.B) {
+	listener, err := net.ListenPacket(statsdProtocol, "localhost:")
+	if err != nil {
```
Comment (azdagron) [inline pkg/common/telemetry/metrics_benchmark_test.go]: nit: `if err != nil` check is not required (performed internally by `require.NoError`)

There are a bunch of similar checks throughout the file.

**Comment by amoore877** [comment]
> The statsd sink writes udp packets to localhost 8125 (the configured stats server). Since you don't have anything bound to that port we log that the metric wasn't able to be written. Maybe `testStatsdConfig` can take a port number (to init the config) and in your benchmarks, you can start up a udp listener on an ephemeral port (that you pass to `testStatsdConfig`) that just discards incoming packets?

I've added that, though strangely I still get some errors:

```
BenchmarkStatsd/SetGauge-8         	2019/11/25 15:42:59 [ERR] Error writing to statsd! Err: write udp [::1]:49385->[::1]:49384: write: connection refused
  870622	      1191 ns/op	     340 B/op	       8 allocs/op
BenchmarkStatsd/SetGaugeWithLabels-8         	  379404	      3018 ns/op	    1074 B/op	      24 allocs/op
BenchmarkStatsd/EmitKey-8                    	 1268426	       950 ns/op	     228 B/op	       6 allocs/op
BenchmarkStatsd/IncrCounter-8                	2019/11/25 15:43:04 [ERR] Error writing to statsd! Err: write udp [::1]:58517->[::1]:49384: write: connection refused
  982852	      1186 ns/op	     340 B/op	       8 allocs/op
BenchmarkStatsd/IncrCounterWithLabels-8      	  384928	      3068 ns/op	    1075 B/op	      24 allocs/op
BenchmarkStatsd/AddSample-8                  	 1000000	      1192 ns/op	     340 B/op	       8 allocs/op
BenchmarkStatsd/AddSampleWithLabels-8        	  391896	      3055 ns/op	    1075 B/op	      24 allocs/op
BenchmarkStatsd/MeasureSince-8               	2019/11/25 15:43:09 [ERR] Error writing to statsd! Err: write udp [::1]:58518->[::1]:49384: write: connection refused
  962360	      1343 ns/op	     340 B/op	       8 allocs/op
BenchmarkStatsd/MeasureSinceWithLabels-8     	  359241	      3286 ns/op	    1076 B/op	      24 allocs/op
```

File: pkg/common/telemetry/statsd_test.go [hunk 3]
```diff
@@ -52,7 +58,7 @@ func TestStatsdRun(t *testing.T) {
 	}
 }
 
-func testStatsdConfig() *MetricsConfig {
+func testStatsdConfig(port int) *MetricsConfig {
```
Comment (azdagron) [inline pkg/common/telemetry/statsd_test.go]: maybe have a `testdStatsConfig` that uses the default test port and a `testStatsdConfigWithPort` that takes the port? that would cut down on having to pass the default port into every `testStatsdConfig` callsite.

**Comment by azdagron** [comment]
Did some packet capture and this only happens when the sink writes packets to the ipv6 loopback address (i.e. `::1`) when the server is listening on ipv4 loopback (and vice-versa). It must be due to how "localhost" is resolved. If both sides use `127.0.0.1` then everything works.

**Comment by amoore877** [comment]
> Did some packet capture and this only happens when the sink writes packets to the ipv6 loopback address (i.e. `::1`) when the server is listening on ipv4 loopback (and vice-versa). It must be due to how "localhost" is resolved. If both sides use `127.0.0.1` then everything works.

works now :) thanks for looking into that

**Comment by amoore877** [comment]
latest output:
```
BenchmarkDogStatsd/SetGauge-8         	  249458	      4677 ns/op	     384 B/op	      10 allocs/op
BenchmarkDogStatsd/SetGaugeWithLabels-8         	  177241	      7295 ns/op	    1115 B/op	      33 allocs/op
BenchmarkDogStatsd/EmitKey-8                    	 4863430	       318 ns/op	     144 B/op	       2 allocs/op
BenchmarkDogStatsd/IncrCounter-8                	  288457	      4040 ns/op	     384 B/op	      10 allocs/op
BenchmarkDogStatsd/IncrCounterWithLabels-8      	  184268	      6655 ns/op	    1096 B/op	      33 allocs/op
BenchmarkDogStatsd/AddSample-8                  	  284901	      4169 ns/op	     384 B/op	      10 allocs/op
BenchmarkDogStatsd/AddSampleWithLabels-8        	  176631	      6839 ns/op	    1112 B/op	      33 allocs/op
BenchmarkDogStatsd/MeasureSince-8               	  264140	      4665 ns/op	     400 B/op	      10 allocs/op
BenchmarkDogStatsd/MeasureSinceWithLabels-8     	  167151	      7478 ns/op	    1112 B/op	      33 allocs/op
BenchmarkInMem/SetGauge-8                       	  652495	      1760 ns/op	     921 B/op	      22 allocs/op
BenchmarkInMem/SetGaugeWithLabels-8             	  293972	      4133 ns/op	    1761 B/op	      46 allocs/op
BenchmarkInMem/EmitKey-8                        	  925632	      1305 ns/op	     687 B/op	      15 allocs/op
BenchmarkInMem/IncrCounter-8                    	  686224	      1788 ns/op	     920 B/op	      22 allocs/op
BenchmarkInMem/IncrCounterWithLabels-8          	  294072	      4082 ns/op	    1733 B/op	      46 allocs/op
BenchmarkInMem/AddSample-8                      	  649711	      1827 ns/op	     920 B/op	      22 allocs/op
BenchmarkInMem/AddSampleWithLabels-8            	  307620	      4003 ns/op	    1733 B/op	      46 allocs/op
BenchmarkInMem/MeasureSince-8                   	  647295	      1838 ns/op	     920 B/op	      22 allocs/op
BenchmarkInMem/MeasureSinceWithLabels-8         	  311511	      4005 ns/op	    1733 B/op	      46 allocs/op
BenchmarkM3/SetGauge-8                          	  949958	      1256 ns/op	     992 B/op	      12 allocs/op
BenchmarkM3/SetGaugeWithLabels-8                	  375645	      3117 ns/op	    1515 B/op	      27 allocs/op
BenchmarkM3/EmitKey-8                           	 4947486	       242 ns/op	     144 B/op	       2 allocs/op
BenchmarkM3/IncrCounter-8                       	  977611	      1254 ns/op	     992 B/op	      12 allocs/op
BenchmarkM3/IncrCounterWithLabels-8             	  384201	      3083 ns/op	    1515 B/op	      27 allocs/op
BenchmarkM3/AddSample-8                         	  984825	      1266 ns/op	    1024 B/op	      13 allocs/op
BenchmarkM3/AddSampleWithLabels-8               	  367219	      3165 ns/op	    1547 B/op	      28 allocs/op
BenchmarkM3/MeasureSince-8                      	  922942	      1405 ns/op	    1024 B/op	      13 allocs/op
BenchmarkM3/MeasureSinceWithLabels-8            	  371704	      3231 ns/op	    1546 B/op	      28 allocs/op
BenchmarkPrometheus/SetGauge-8                  	  798486	      1511 ns/op	     467 B/op	      12 allocs/op
BenchmarkPrometheus/SetGaugeWithLabels-8        	  331149	      3586 ns/op	    1143 B/op	      35 allocs/op
BenchmarkPrometheus/EmitKey-8                   	 5060436	       235 ns/op	     144 B/op	       2 allocs/op
BenchmarkPrometheus/IncrCounter-8               	  848130	      1492 ns/op	     467 B/op	      12 allocs/op
BenchmarkPrometheus/IncrCounterWithLabels-8     	  321237	      3582 ns/op	    1145 B/op	      35 allocs/op
BenchmarkPrometheus/AddSample-8                 	  789154	      1581 ns/op	     467 B/op	      12 allocs/op
BenchmarkPrometheus/AddSampleWithLabels-8       	  336206	      3596 ns/op	    1144 B/op	      35 allocs/op
BenchmarkPrometheus/MeasureSince-8              	  706092	      1609 ns/op	     467 B/op	      12 allocs/op
BenchmarkPrometheus/MeasureSinceWithLabels-8    	  333679	      3717 ns/op	    1144 B/op	      35 allocs/op
BenchmarkStatsd/SetGauge-8                      	 1000000	      1253 ns/op	     340 B/op	       8 allocs/op
BenchmarkStatsd/SetGaugeWithLabels-8            	  343230	      3134 ns/op	    1057 B/op	      24 allocs/op
BenchmarkStatsd/EmitKey-8                       	 1221610	       971 ns/op	     228 B/op	       6 allocs/op
BenchmarkStatsd/IncrCounter-8                   	 1000000	      1213 ns/op	     340 B/op	       8 allocs/op
BenchmarkStatsd/IncrCounterWithLabels-8         	  393278	      3341 ns/op	    1057 B/op	      24 allocs/op
BenchmarkStatsd/AddSample-8                     	  869545	      1302 ns/op	     340 B/op	       8 allocs/op
BenchmarkStatsd/AddSampleWithLabels-8           	  372062	      3258 ns/op	    1056 B/op	      24 allocs/op
BenchmarkStatsd/MeasureSince-8                  	  916932	      1347 ns/op	     340 B/op	       8 allocs/op
BenchmarkStatsd/MeasureSinceWithLabels-8        	  361531	      3300 ns/op	    1055 B/op	      24 allocs/op
```

**Comment by azdagron** [comment]
Hi @amoore877! Thanks for putting these benchmarks together. It was good validation that the sinks are roughly on-par. In light of #1248 and the eventual transition towards disabling of the in-mem sink by default, do we still think these benchmarks have high value living in the codebase?

**Comment by amoore877** [comment]
> Hi @amoore877! Thanks for putting these benchmarks together. It was good validation that the sinks are roughly on-par. In light of #1248 and the eventual transition towards disabling of the in-mem sink by default, do we still think these benchmarks have high value living in the codebase?

I think it may be good to keep so that we can compare against any future metrics changes. Ideally I'd like to in the future also add benchmarking against logging, crypto usage, and DB accesses. These are the operations spire has the most calls against so getting us started on measuring how efficiently we're actually coding, and how efficient our third party libraries are, would be beneficial.

Is there a way to get Travis CI to run benchmarking and do comparisons against HEAD vs. PR? It would be great to be able to have automated detection of when we are hurting performance with changes.

**Comment by azdagron** [comment]
We could have travis run the benchmarks, output the results, and use something like (benchstat)[https://godoc.org/golang.org/x/perf/cmd/benchstat] to determine if there was a significant change. But none of that is wired up currently.

**Comment by amoore877** [comment]
> We could have travis run the benchmarks, output the results, and use something like (benchstat)[https://godoc.org/golang.org/x/perf/cmd/benchstat] to determine if there was a significant change. But none of that is wired up currently.

A future enhancement? :)

**Pull request merged**